book/html.md

```diff
@@ -5,408 +5,559 @@ prev: text
+can be omitted, like in `<input disabled>`. In this case, the
+attribute value is supposed to default to the empty string:
 
-``` {.python}
-while current.parent:
-    current = current.parent
+``` {.python indent=8}
+for attrpair in parts[1:]:
+    if "=" in attrpair:
+        # ...
+    else:
+        self.attributes[attrpair.lower()] = ""
 ```
 
-These rules for malformed HTML may seem arbitrary, and they are. But
-they evolved over years of trying to guess what people \"meant\" when
-they wrote that HTML, and are now codified in the [HTML 5 parsing
-algorithm](https://www.w3.org/TR/2011/WD-html5-20110113/parsing.html),
-which spells out in detail how to handle user errors. The tweaks above
-are much more limited, but give you some sense of what the full
-algorithm is like.
+Finally, this code misbehaves on quoted values, since it includes the
+quotes as part of the value. We can fix that:
 
+``` {.python indent=12}
+if "=" in attrpair:
+    if len(value) > 2 and value[0] in ["'", "\""]:
+        value = value[1:-1]
+    # ...
+```
 
-HTML attributes
-===============
+This conditional checks the first character of the value to determine
+if it's quoted, and if so strips off the first and last character,
+leaving the contents of the quotes.
 
-HTML tags have tag names but also *attributes*, which are added to an
-open tag to give additional information about that element. For example,
-the `lang` attribute tells you what language the text in that element is
-in, while the `href` attribute on `a` tags gives the URL that the link
-points to.[^5] Attributes look like this:
+::: {.further}
+This is [not the right way][case-hard] to do case
+insensitive comparisons; the Unicode case folding algorithm should be
+used if you want to handle languages other than English. But in HTML
+specifically, tag names only use the ASCII characters where this test
+is sufficient.
+:::
+    
+[case-hard]: https://www.b-list.org/weblog/2018/nov/26/case/
 
-``` {.example}
-<tagname attrname=attrvalue attrname=attrvalue>
+
+Doctype declarations
+====================
+
+Self-closing elements are now handled correctly and the list of open
+elements when parsing this page is now shorter:
+
+```
+[<!DOCTYPE>]
 ```
 
-You can have any number of attributes in an open tag; the names are
-case-insensitive, while the values are an arbitrary string of character
-that can even include spaces if you surround them with quotes. Also, the
-values are optional; \"`<tagname attrname>`\" is perfectly valid and
-sets that attribute to an empty string.
+[html5-doctype]: https://html.spec.whatwg.org/multipage/syntax.html#the-doctype
 
-Let\'s extend `ElementNode` to store attributes:
+This special tag is called a [doctype][html5-doctype], and it's not
+really an element at all. It's not supposed to have close tag, but we
+can't mark it a self-closing tag—it's always the very first thing in
+the document, so there wouldn't be an open element to append it to.
+It's best to throw it away:
```

Comment:
```suggestion
It's best to throw it away for now (a real browser needs to parse it to handle various legacy or obscure parsing modes[1]):
```
[1] https://en.wikipedia.org/wiki/Quirks_mode

Reply:
Oh I see you refer to quirks mode below. Maybe just foreshadow in a parenthetical?

Reply:
Worth footnoting.

Reply:
You know, looking over the HTML5 spec again, maybe the right move here is to only drop whitespace text. HTML5 mandates that in standards mode (https://html.spec.whatwg.org/multipage/parsing.html#the-before-html-insertion-mode). And then implicit tags later in the chapter handle the case of text being inserted here.

Reply:
Oh, sorry, I got the context wrong here. I'll add a footnote.